File: Stack/844. Backspace String Compare/Solution.py
'''
    时间复杂度 O(n)
    空间复杂度O(n)

    利用数组模拟栈解决
'''
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
            def toString(S):
                ans = []
                for c in S:
                    if c == "#":
                        ans and ans.pop()
                    else:
                        ans.append(c)
                return ''.join(ans)
            return toString(S) == toString(T)

File: Stack/844. Backspace String Compare/Solution.java
/*
    时间复杂度 O(n)
    空间复杂度O(n)

    利用栈解决
*/
class Solution {
    public boolean backspaceCompare(String S, String T) {
        
        return toString(S).equals(toString(T));
    }
    
    public Stack toString(String str){
        
        Stack stack = new Stack();
        for(for char c : str.toCharArray()){
            if(c == '#'){
                if(!stack.isEmpty()){
                    stack.pop();
                }
            }else{
                stack.push(c);
            }
        }
        return stack;
    }
}